Add extra context to config-key validation error message

wjordan opened the pull request:
# Description

Make the config-key validation error more helpful with a more detailed error message including the source filename where invalid config keys are being referenced and a reference to further documentation on the configuration system.

Before:

```
Unknown property not in defaults: foo (RuntimeError)
```

After:

```
Invalid configuration (Cdo::Config::UnknownKeyError)
Remove unknown keys not defined in ./config.yml.erb:
foo (in ./locals.yml)
bar (in ./locals.yml)
Refer to ./config/README.md for further documentation.
```

## Testing story

Unit test covering additional behavior included in the PR.

uponthesun commented:
Should be `./config/README.md` instead of `/config/README.md`, otherwise looks great!

wjordan commented:
Updated the error message path, and reworked it a bit so that all of the path names hyperlink in RubyMine:
![image](https://user-images.githubusercontent.com/56541/66868684-a2407200-ef52-11e9-9306-baf29d775870.png)

wjordan commented:
Strange CI error, `commit_hash` file not found from `build:apps` task:

```
rake aborted!
--
Errno::ENOENT: No such file or directory @ rb_sysopen - ./apps/build/commit_hash
/drone/src/lib/rake/build.rake:29:in `write'
/drone/src/lib/rake/build.rake:29:in `block (3 levels) in <top (required)>'
/drone/src/lib/rake/build.rake:9:in `chdir'
/drone/src/lib/rake/build.rake:9:in `block (2 levels) in <top (required)>'
[...]
/drone/src/lib/rake/build.rake:146:in `block (2 levels) in <top (required)>'
/drone/src/lib/cdo/chat_client.rb:62:in `wrap'
/drone/src/lib/rake/build.rake:146:in `block in <top (required)>'
[...]
Tasks: TOP => build:all => build:apps
[2019-10-16 00:14:39] [test] <b>build</b> failed in 5:45 minutes
```
